docs/essentials/gc/unmanaged/implementingdispose.md

```diff
@@ -1 +1,442 @@
+         throw new FileNotFoundException(String.Format("Cannot open '{0}'", filename));
+
+      // Get file size.
+      bufferSize = GetFileSize(safeHandle, out upperWord); 
+      if (bufferSize == INVALID_FILE_SIZE)
+         bufferSize = -1;
+      else if (upperWord > 0) 
+         bufferSize = (((long)upperWord) << 32) + bufferSize;
+   }
+
+   public long Size 
+   { get { return bufferSize; } }
+
+   public void Dispose()
+   {
+      Dispose(true);
+      GC.SuppressFinalize(this);
+   }           
+
+   protected virtual void Dispose(bool disposing)
+   {
+      if (disposed) return;
+
+      // Dispose of managed resources here.
+      if (disposing)
+         safeHandle.Dispose();
+
+      // Dispose of any unmanaged resources not wrapped in safe handles.
+
+      disposed = true;
+   }  
+}
+```
+
+## Using a safe handle to implement the dispose pattern for a derived class
+
+The following example illustrates the dispose pattern for a derived class, `DisposableStreamResource2`, that inherits from the `DisposableStreamResource` class presented in the previous example. The class adds an additional method, `WriteFileInfo`, and uses a `SafeFileHandle` object to wrap the handle of the writable file.
+
+```csharp
+using Microsoft.Win32.SafeHandles;
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+public class DisposableStreamResource2 : DisposableStreamResource
+{
+   // Define additional constants.
+   protected const uint GENERIC_WRITE = 0x40000000; 
+   protected const uint OPEN_ALWAYS = 4;
+
+   // Define additional APIs.
+   [DllImport("kernel32.dll")]   
+   protected static extern bool WriteFile(
+                                SafeFileHandle safeHandle, string lpBuffer, 
+                                int nNumberOfBytesToWrite, out int lpNumberOfBytesWritten,
+                                IntPtr lpOverlapped);
+
+   // Define locals.
+   private bool disposed = false;
+   private string filename;
+   private bool created = false;
+   private SafeFileHandle safeHandle;
+
+   public DisposableStreamResource2(string filename) : base(filename)
+   {
+      this.filename = filename;
+   }
+
+   public void WriteFileInfo()
+   { 
+      if (! created) {
+         IntPtr hFile = CreateFile(@".\FileInfo.txt", GENERIC_WRITE, 0, 
+                                   IntPtr.Zero, OPEN_ALWAYS, 
+                                   FILE_ATTRIBUTE_NORMAL, IntPtr.Zero);
+         if (hFile != INVALID_HANDLE_VALUE)
+            safeHandle = new SafeFileHandle(hFile, true);
+         else
+            throw new IOException("Unable to create output file.");
+
+         created = true;
+      }
+
+      string output = String.Format("{0}: {1:N0} bytes\n", filename, Size);
+      int bytesWritten;
+      bool result = WriteFile(safeHandle, output, output.Length, out bytesWritten, IntPtr.Zero);                                     
+   }
+
+   protected new virtual void Dispose(bool disposing)
```

Comment:
:exclamation: This has to be `override`, not `new virtual`. The way it is written, `DisposableStreamResource2.Dispose(bool)` will never be called, so `DisposableStreamResource2.safeHandle` will never be disposed.